external/js/kyber/lib/sign/schnorr/schnorr.js

```diff
@@ -0,0 +1,127 @@
+const group = require("./../../group");
+// XXX to be changed to an interface style
+const nist = group.nist;
+const curve = nist.Curve;
+const scalarCons = nist.Scalar;
+const pointCons = nist.Point;
+
+const hash = require("hash.js");
+/**
+ *
+ * Sign computes a Schnorr signature over the given message.
+ *
+ * @param privateKey private key scalar to sign with
+ * @param message message over which the signature is computed
+ * @return signature as a Uint8Array
+ * */
+function Sign(suite, privateKey, message) {
+  if (suite.constructor !== curve) {
+    throw "first argument must be a suite";
+  }
+  if (privateKey.constructor !== scalarCons) {
+    throw "second argument must be a scalar";
+  }
+  if (message.constructor !== Uint8Array) {
+    throw "third argument must be Uint8Array";
+  }
+
+  // generate r & R
+  const r = suite.scalar().pick();
+  const R = suite.point().mul(r, null);
+  const buffR = R.marshalBinary();
+
+  // generate public key
+  const pub = suite.point().mul(privateKey, null);
```

Comment:
This is surprising, I would have expected that the private and public keys arrive together in a keypair object, so that the public does not need to be derived on each call. Then I went to the Go implementation and was also surprised to see the same thing there.

So: I'm not asking you to change this, but why is it this way?

Reply:
I don't have a definite answer but I would say that it's much safer and simpler this way, even for a little performance penalty. If you give a public key as argument, you should make sure it's the corresponding public key from the secret key otherwise your signature will be wrong. To do that check, you anyway need to do the scalar base multiplication.